src/Frame.cpp

```diff
@@ -1,227 +1,493 @@
-                );
-            }
-            auto d = distance(i, j);
-            auto radii = i_radius.value() + j_radius.value();
-            if (0.03 < d && d < 0.6 * radii && d < cutoff) {
-                topology_.add_bond(i, j);
-            }
-        }
-    }
-
-    auto bonds = topology().bonds();
-    auto to_remove = std::vector<Bond>();
-    // We need to remove bonds between hydrogen atoms which are bonded more than
-    // once
-    for (auto& bond : bonds) {
-        auto i = bond[0], j = bond[1];
-        if (topology_[i].type() != "H") {
-            continue;
-        }
-        if (topology_[j].type() != "H") {
-            continue;
-        }
-
-        auto nbonds = std::count_if(
-            std::begin(bonds), std::end(bonds), [=](const Bond& b) {
-                return b[0] == i || b[0] == j || b[1] == i || b[1] == j;
-            });
-        assert(nbonds >= 1);
-
-        if (nbonds != 1) {
-            to_remove.push_back(bond);
-        }
-    }
-
-    for (auto& bond : to_remove) {
-        topology_.remove_bond(bond[0], bond[1]);
-    }
+         }
+         auto d = distance(i, j);
+         auto radii = i_radius.value() + j_radius.value();
+         if (0.03 < d && d < 0.6 * radii && d < cutoff) {
+            topology_.add_bond(i, j);
+         }
+      }
+   }
+
+   auto bonds = topology().bonds();
+   auto to_remove = std::vector<Bond>();
+   // We need to remove bonds between hydrogen atoms which are bonded more than
+   // once
+   for (auto& bond : bonds) {
+      auto i = bond[0], j = bond[1];
+      if (topology_[i].type() != "H") {
+         continue;
+      }
+      if (topology_[j].type() != "H") {
+         continue;
+      }
+
+      auto nbonds = std::count_if(
+         std::begin(bonds), std::end(bonds), [=](const Bond& b) {
+            return b[0] == i || b[0] == j || b[1] == i || b[1] == j;
+         });
+      assert(nbonds >= 1);
+
+      if (nbonds != 1) {
+         to_remove.push_back(bond);
+      }
+   }
+
+   for (auto& bond : to_remove) {
+      topology_.remove_bond(bond[0], bond[1]);
+   }
+}
+
+void Frame::getMinMaxBox(std::vector<Vector3D> pos, int n, Vector3D& min, Vector3D& max) {
+
+   float x1, x2, y1, y2, z1, z2;
+
+   //Consider first atom's coordinates as min and max.
+   x1 = pos[0][0];
+   x2 = pos[0][0];
+   y1 = pos[0][1];
+   y2 = pos[0][1];
+   z1 = pos[0][2];
+   z2 = pos[0][2];
+
+   for (int i = 1; i < n; i++) {
+      if (pos[i][0] < x1) x1 = pos[i][0];
+      if (pos[i][0] > x2) x2 = pos[i][0];
+      if (pos[i][1] < y1) y1 = pos[i][1];
+      if (pos[i][1] > y2) y2 = pos[i][1];
+      if (pos[i][2] < z1) z1 = pos[i][2];
+      if (pos[i][2] > z2) z2 = pos[i][2];
+   }
+
+   min[0] = x1;
+   max[0] = x2;
+   min[1] = y1;
+   max[1] = y2;
+   min[2] = z1;
+   max[2] = z2;
+}
+
+int Frame::createNeighborList(std::vector<std::vector<int>>& nbList, int xCells, int yCells, int zCells) {
```

Comment:
`std::vector<std::vector<int>>` have horrible CPU cache behavior (cf https://stackoverflow.com/q/38244435/4692076), this might explain part of the slowdown you see.

Reply:
You are right, I didn't check the performance of such data structure using `std::vector`. For the `nblist` variable (known size of sub-vector), I am changing this to `std::vector<std::array<int, 27>>`. From first runs, this modification led to a high decrease of the slowdown seen before.